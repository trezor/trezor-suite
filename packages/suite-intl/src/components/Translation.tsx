import React from 'react';
import { FormattedMessage, MessageDescriptor } from 'react-intl';
import { messages } from '@suite/messages';

export type TranslationProps = {
    isNested?: boolean;
    isReactNative?: boolean; // used to render span on the web, plain node for react-native
    translationTooltip?: React.ComponentType<TooltipProps>;
} & ExtendedMessageDescriptor;

export interface TooltipProps {
    isNested?: boolean;
    messageId: keyof typeof messages;
    children: any;
}

type PrimitiveType = string | number | boolean | Date | null | undefined;

// Add MessageDescriptor type to values entry
type FormatXMLElementFn = (...args: any[]) => string | object;
export interface ExtendedMessageDescriptor extends MessageDescriptor {
    id: keyof typeof messages;
    values?: {
        [key: string]:
            | PrimitiveType
            | React.ReactElement
            | ExtendedMessageDescriptor
            | FormatXMLElementFn;
    };
}

export const isMsgType = (props: TranslationProps | React.ReactNode): props is TranslationProps =>
    typeof props === 'object' && props !== null && (props as TranslationProps).id !== undefined;

const RenderChildrenComponent = ({ children }: any) => children;

export const Translation = (props: TranslationProps) => {
    // Tooltip is only used in suite-landing-page, in case landing will be moved to eshop this could be removed
    const TooltipComponent = props.translationTooltip ?? RenderChildrenComponent;
    const values: Record<string, PrimitiveType | React.ReactNode | ExtendedMessageDescriptor> = {};
    // message passed via props (id, defaultMessage, values)
    Object.keys(props.values || []).forEach(key => {
        // Iterates through all values. The entry may also contain a MessageDescriptor.
        // If so, Renders MessageDescriptor by passing it to `Translation` component
        const maybeMsg = props.values![key];
        values[key] = isMsgType(maybeMsg) ? (
            <Translation {...maybeMsg} translationTooltip={props.translationTooltip} isNested />
        ) : (
            maybeMsg
        );
    });

    // prevent runtime errors
    if (
        !props.defaultMessage &&
        Object.prototype.hasOwnProperty.call(props, 'id') &&
        !messages[props.id]
    ) {
        return <>{`Unknown translation id: ${props.id}`}</>;
    }

    const defaultTagName = props.isReactNative || props.isNested ? undefined : 'span';

    return (
        <TooltipComponent isNested={props.isNested} messageId={props.id}>
            <FormattedMessage
                id={props.id}
                tagName={defaultTagName}
                defaultMessage={props.defaultMessage || messages[props.id].defaultMessage}
                // pass undefined to a 'values' prop in case of an empty values object
                values={Object.keys(values).length === 0 ? undefined : values}
            />
        </TooltipComponent>
    );
};
