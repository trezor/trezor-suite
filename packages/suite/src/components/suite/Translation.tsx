import { ReactNode } from 'react';
import { FormattedMessage } from 'react-intl';

import {
    ExtendedMessageDescriptor as CommonExtendedMessageDescriptor,
    FormatXMLElementFn,
} from '@suite-common/intl-types';

// We cannot use aliases here because this file is directly imported by the @suite-common/intl-types
// It's little hacky by this will be solved when PR for refactor intl will be merged.
import messages from '../../support/messages';

export type TranslationKey = keyof typeof messages;

type OwnProps = {
    isNested?: boolean;
};

export type ExtendedMessageDescriptor = CommonExtendedMessageDescriptor;
type MsgType = OwnProps & ExtendedMessageDescriptor;

export const isMsgType = (
    props: MsgType | ReactNode | ExtendedMessageDescriptor | Date | FormatXMLElementFn,
): props is MsgType =>
    typeof props === 'object' && props !== null && (props as MsgType).id !== undefined;

export const Translation = (props: MsgType) => {
    const values: Record<string, any> = {};
    // message passed via props (id, defaultMessage, values)
    Object.keys(props.values || []).forEach(key => {
        // Iterates through all values. The entry may also contain a MessageDescriptor.
        // If so, Renders MessageDescriptor by passing it to `Translation` component
        const maybeMsg = props.values![key];
        values[key] = isMsgType(maybeMsg) ? <Translation {...maybeMsg} isNested /> : maybeMsg;
    });

    // prevent runtime errors
    if (
        !props.defaultMessage &&
        Object.prototype.hasOwnProperty.call(props, 'id') &&
        !messages[props.id]
    ) {
        return <>{`Unknown translation id: ${props.id}`}</>;
    }

    const defaultTagName = props.isNested ? undefined : 'span';

    return (
        <FormattedMessage
            id={props.id}
            tagName={defaultTagName}
            defaultMessage={props.defaultMessage || messages[props.id].defaultMessage}
            // pass undefined to a 'values' prop in case of an empty values object
            values={Object.keys(values).length === 0 ? undefined : values}
        />
    );
};
