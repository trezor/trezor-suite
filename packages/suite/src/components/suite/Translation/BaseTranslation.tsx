import { Translation } from '@suite/intl';

/**
 * @deprecated Use `Translation` from `@suite/intl` instead.
 */
const BaseTranslation = Translation;

export default BaseTranslation;
