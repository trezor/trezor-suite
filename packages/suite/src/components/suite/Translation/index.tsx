import React from 'react';
import { FormattedMessage } from 'react-intl';
import { ExtendedMessageDescriptor } from '@suite-types';

export const isChildrenMessageDescriptor = (
    message: null | React.ReactNode | ExtendedMessageDescriptor,
): message is ExtendedMessageDescriptor => {
    return (
        typeof message === 'object' &&
        message !== null &&
        (message as ExtendedMessageDescriptor).defaultMessage !== undefined
    );
};

export const isMsgType = (props: MsgType | ChildrenType): props is MsgType => {
    return (
        typeof props === 'object' &&
        props !== null &&
        (props as MsgType).defaultMessage !== undefined
    );
};

export const isChildrenType = (props: MsgType | ChildrenType): props is ChildrenType => {
    return typeof props === 'object' && props !== null && props.children !== null;
};

interface ChildrenType {
    children: React.ReactNode | ExtendedMessageDescriptor;
}

interface MsgType extends ExtendedMessageDescriptor {
    children?: undefined;
}

type PrimitiveType = string | number | boolean | Date | null | undefined;

/**
 * Util component that helps with rendering react-intl messages.
 * It extends the FormattedMessage API and adds prop 'children' (ReactNode or MessageDescriptor).
 * The standard way of using it is by passing all necessary props (id, defaultMessage) to the component (eg. by spreading message obj)
 * The alternative is to pass whole (Extended)MessageDescriptor object as a children.
 * If children prop is an object, it assumes that it is ExtendedMessageDescriptor
 * and renders FormattedMessage with the message. Otherwise it just simply render passed children.
 */
const Translation = (props: ChildrenType | MsgType) => {
    const values: Record<string, PrimitiveType | React.ReactNode | ExtendedMessageDescriptor> = {};
    // message passed via props (id, defaultMessage, values)
    if (isMsgType(props)) {
        Object.keys(props.values || []).forEach(key => {
            // Iterates through all values. The entry may also contain a MessageDescriptor.
            // If so, Renders MessageDescriptor by passing it to `Translation` component
            values[key] = isChildrenMessageDescriptor(props.values![key]) ? (
                <Translation>{props.values![key]}</Translation>
            ) : (
                props.values![key]
            );
        });

        // pass undefined to a 'values' prop in case of an empty values object
        return (
            <FormattedMessage
                {...props}
                values={Object.keys(values).length === 0 ? undefined : values}
            />
        );
    }
    // Passed children prop (ExtendedMessageDescriptor obj)
    // Basically same logic as for MsgType
    if (isChildrenType(props) && isChildrenMessageDescriptor(props.children)) {
        Object.keys(props.children.values || []).forEach(key => {
            // @ts-ignore
            const val = props.children.values[key];
            values[key] = isChildrenMessageDescriptor(val) ? <Translation>{val}</Translation> : val;
        });

        return (
            <FormattedMessage
                {...props.children}
                values={Object.keys(values).length === 0 ? undefined : values}
            />
        );
    }

    // passed children prop (regular ReactChild, not ExtendedMessageDescriptor)
    return <>{props.children}</>;
};

export { Translation };
