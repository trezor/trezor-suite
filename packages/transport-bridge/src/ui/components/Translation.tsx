import { FormattedMessage } from 'react-intl';

import type { Messages } from '../translations/default';

export const Translation = ({ id, values }: { id: keyof Messages; values?: any }) => {
    return <FormattedMessage id={id} tagName="span" values={values} />;
};
