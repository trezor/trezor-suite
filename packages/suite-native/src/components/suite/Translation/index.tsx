import React from 'react';
import BaseTranslation from '@suite/components/suite/Translation/components/BaseTranslation';

type TranslationProps = Omit<React.ComponentProps<typeof BaseTranslation>, 'translationTooltip'>;
const Translation = (props: TranslationProps) => {
    return (
        <BaseTranslation
            {...props}
            isReactNative
            translationTooltip={({ children }) => {
                return children;
            }}
        />
    );
};

export default Translation;
