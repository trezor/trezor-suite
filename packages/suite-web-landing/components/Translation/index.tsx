import React, { useContext } from 'react';
import BaseTranslation from '@suite-components/Translation/components/BaseTranslation';
import HelperTooltip, {
    Props as HelperTooltipProps,
} from '@suite-components/Translation/components/HelperTooltip';

export const TranslationModeContext = React.createContext(false);

const CustomHelperTooltip = (props: HelperTooltipProps) => {
    const translationMode = useContext(TranslationModeContext);

    return (
        <HelperTooltip {...props} language="en" translationMode={translationMode}>
            {props.children}
        </HelperTooltip>
    );
};

type TranslationProps = Omit<React.ComponentProps<typeof BaseTranslation>, 'translationTooltip'>;
const Translation = (props: TranslationProps) => {
    return <BaseTranslation {...props} translationTooltip={CustomHelperTooltip} />;
};

export default Translation;
