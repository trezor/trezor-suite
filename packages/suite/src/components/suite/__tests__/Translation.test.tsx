import React from 'react';
import createComponentWithIntl from '../../../support/tests/IntlHelper';
import { Translation } from '../Translation';

const messages = {
    TR_HELLO: {
        id: 'TR_HELLO',
        defaultMessage: 'Hello',
        descriptor: 'test desc',
    },
    TR_HELLO_NAME: {
        id: 'TR_HELLO_NAME',
        defaultMessage: 'Hello {TR_NAME}, {TR_AGE}',
    },
    TR_NAME: {
        id: 'TR_NAME',
        defaultMessage: 'Name: {name}',
    },
    TR_AGE: {
        id: 'TR_NAME',
        defaultMessage: 'Age: {age}',
    },
};

describe('Translation component', () => {
    test('with id, defaultMessage props', () => {
        const component = createComponentWithIntl(<Translation id="TR_CANCEL" />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds string in value (passed via props)', () => {
        const component = createComponentWithIntl(
            // @ts-ignore
            <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another message object in values (passed via props)', () => {
        const component = createComponentWithIntl(
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    // @ts-ignore
                    TR_NAME: { ...messages.TR_NAME, values: { name: 'John' } },
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('with message that holds another in values (passed via props)', () => {
        const component = createComponentWithIntl(
            // @ts-ignore
            <Translation
                {...messages.TR_HELLO_NAME}
                values={{
                    // @ts-ignore
                    TR_NAME: <Translation {...messages.TR_NAME} values={{ name: 'John' }} />,
                    TR_AGE: 100,
                }}
            />,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
