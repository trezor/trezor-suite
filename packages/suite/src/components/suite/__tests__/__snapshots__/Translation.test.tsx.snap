// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation component with ReactNode as a children 1`] = `
<div>
  test
</div>
`;

exports[`Translation component with id, defaultMessage props 1`] = `"Hello"`;

exports[`Translation component with message obj as children 1`] = `"Hello"`;

exports[`Translation component with message that holds another in values (passed via props) 1`] = `
Array [
  "Hello ",
  "Name: ",
  "John",
  ", ",
  "100",
]
`;

exports[`Translation component with message that holds another message object in values (passed via children) 1`] = `
Array [
  "Hello ",
  "Name: ",
  "John",
  ", ",
  "100",
]
`;

exports[`Translation component with message that holds another message object in values (passed via props) 1`] = `
Array [
  "Hello ",
  "Name: ",
  "John",
  ", ",
  "100",
]
`;

exports[`Translation component with message that holds string in value (passed as children) 1`] = `
Array [
  "Name: ",
  "John",
]
`;

exports[`Translation component with message that holds string in value (passed via props) 1`] = `
Array [
  "Name: ",
  "John",
]
`;

exports[`Translation component with string as a children 1`] = `"test"`;
